### DISCUSSION TITLE
How can I set variable within location?

### DISCUSSION THREAD

**Discussion by huwylphimet** [nginx/ansible-role-nginx-config#410] [Q&A] [answered]
Hi,
How can I set variable within location (see @handle_redirects in the example below)?
In the [template file](https://github.com/nginxinc/ansible-role-nginx-config/blob/2651dcb6be8d2067c0cc803ccdbf75ced1c07dfc/defaults/main/template.yml#L693) I found the _set_ key but in the comment it says _Not available in the 'http' context_

I would like to achieve something like that:
```
http {
    upstream api {
        server api1:8000;
        server api2:8000;
    }

    server {
        listen 80;

        location / {
            proxy_pass http://api;
            proxy_intercept_errors on;
            error_page 301 302 307 = @handle_redirects;
        }

        location @handle_redirects {
            set $original_location $upstream_http_location;  # how to configure the template.yml for setting variable here?
            proxy_pass $original_location;
        }
    }
}
```
Thanks for any advice

**Comment by huwylphimet** [comment]
Hello,
I found a solution myself by using the [_custom_directives_](https://github.com/nginxinc/ansible-role-nginx-config/blob/2651dcb6be8d2067c0cc803ccdbf75ced1c07dfc/defaults/main/template.yml#L964) key like the following:

```
          - location: /
            proxy:
              pass: "http://api"
              intercept_errors: true
            core:
              error_page:
                - code:
                    - 301
                    - 302
                    - 307
                  uri: "@handle_redirects"
          - location: "@handle_redirects"
            custom_directives:
              - set $original_location $upstream_http_location;
              - proxy_pass $original_location;
```

If someone would have a more elegant way, please share it. If not, at least maybe my solution could be useful to someone.

**Comment by alessfg** [comment, marked as answer]
Hey @huwylphimet!

You should be able to use `set` within the `location` block. When a comment says `Not available in the 'http' context` what it means is that it's not available directly within the `http` block, but it's available elsewhere (`server` and `location` in this case) 😄

**Reply by huwylphimet** [reply]
yes, you're right :) I forgot to defined the parent level _rewrite_...
now I've:
```
          - location: "@handle_redirects"
            rewrite:
              set:
                - variable: $original_location
                  value: $upstream_http_location
            proxy:
              pass: $original_location
```
that generates:
```
server {
    location @handle_redirects {
        proxy_pass $original_location;

        set $original_location $upstream_http_location;
    }
}
```
I just noticed that the directives are not in the same order:
I defined the _set_ directive before the _proxy_pass_ but the result is reversed.
Is it possible to set the order of directives in order to have the variable set before the _proxy_pass_ directive? Or this doesn't matter for ningx at all in this case? (my goal would be to make ningx handle the http redirect 30x from the upstreams, see example in the 1st post)

**Reply by alessfg** [reply]
For most use cases the order of `nginx` directives does not matter. There are some edge cases (mostly involving using `rewrite`) where it will matter, but you should be good here. That being said, the `$upstream_http_location` isn't one of the default `nginx` variables as far as I know, so you might run into some issues if that variable is not set when `nginx` starts. (And if I might add a tiny further suggestion, if the above is not a simplified config you could just `proxy_pass` to `$upstream_http_location` directly.)

**Reply by huwylphimet** [reply]
Thank you very much for this feedback!
I'll check for the `$upstream_http_location` state in our production environment (until now I only tested that config in my local docker environment and I can confirm now that the order doesn't matter in this situation).
This config is likely to evolve in the future and storing `$upstream_http_location` might allow us to consume that information after `proxy_pass` is being called (since as soon as `proxy_pass` is called, the `$upstream_http_location` might be overwritten).
But for now, it is only used there and I will try, as you kindly suggested, to `proxy_pass` to `$upstream_http_location` directly and see if nginx is still doing what is expected. 
Many thanks for your help.
Kind regards
